File: backend/services/bd_service/domain/entities/event.py
from dataclasses import dataclass
from datetime import datetime
from typing import Optional


@dataclass
class Event:
    id: Optional[int]
    year: int
    event_name: str
    description: Optional[str]
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    
    MIN_YEAR = 1900
    MAX_YEAR = 2030
    
    def validate(self) -> bool:
        if not self.year or not (self.MIN_YEAR <= self.year <= self.MAX_YEAR):
            return False
        if not self.event_name or len(self.event_name.strip()) < 2:
            return False
        if len(self.event_name) > 500:
            return False
        return True
    
    def get_decade(self) -> int:
        return (self.year // 10) * 10
    
    def get_era(self) -> str:
        if self.year < 1960:
            return "post_war"
        elif self.year < 1980:
            return "cold_war"
        elif self.year < 1991:
            return "late_soviet"
        elif self.year < 2000:
            return "nineties"
        elif self.year < 2010:
            return "two_thousands"
        elif self.year < 2020:
            return "twenty_tens"
        else:
            return "twenty_twenties"

File: backend/services/bd_service_go/internal/domain/event.go
package domain

import (
    "strings"
    "time"
)

const (
    MinYear = 1900
    MaxYear = 2030
)

type Event struct {
    ID          int        `json:"id"`
    Year        int        `json:"year"`
    EventName   string     `json:"event_name"`
    Description *string    `json:"description,omitempty"`
    CreatedAt   *time.Time `json:"created_at,omitempty"`
    UpdatedAt   *time.Time `json:"updated_at,omitempty"`
}

func (e *Event) Validate() bool {
    if e.Year < MinYear || e.Year > MaxYear {
        return false
    }
    trimmedName := strings.TrimSpace(e.EventName)
    if len(trimmedName) < 2 || len(trimmedName) > 500 {
        return false
    }
    return true
}

func (e *Event) GetDecade() int {
    return (e.Year / 10) * 10
}

func (e *Event) GetEra() string {
    switch {
    case e.Year < 1960:
        return "post_war"
    case e.Year < 1980:
        return "cold_war"
    case e.Year < 1991:
        return "late_soviet"
    case e.Year < 2000:
        return "nineties"
    case e.Year < 2010:
        return "two_thousands"
    case e.Year < 2020:
        return "twenty_tens"
    default:
        return "twenty_twenties"
    }
}